Fixes #1372: Add localization plugin documentation (#1376)
google/automotive-design-compose: docs/_docs/localization.md

---
title: Localization
layout: page
nav_order: 12
---

# Localization

We are developing tools for developers to localize the figma design into different languages more easily.

> The first introduced tool is a figma plugin that generates a strings xml files containing the strings

Reviewer: ```suggestion
The first introduced tool is a figma plugin that generates a strings xml file containing the strings
```

Author: Done. Good catch!

used by the figma design.

## 1. Build the localization plugin

> Localization plugin is still under development and it is only available from building the source code.

Reviewer 2: ```suggestion
The localization plugin is still under development and is only available from building the source code.
```

Author: Done.

can download it from the [Figma downloads page][2].

4. Import the plugin from manifest and you are ready to run the plugin. The file path is:

    ```text
    automotive-design-compose/support-figma/extended-layout-plugin/manifest.json
    ```

    <img src="ImportPluginFromManifest.png">

## 2. Run the localization plugin

Now you can run the localization plugin through:
**Figma Main Menu -> Plugins -> Development**
   **-> Automotive Design For Compose -> Localization Tool -> Generate String Resource**

### 2.1 Options

Currently the plugin provides two options to exclude text nodes from translations.

<img src="LocalizationOptions.png">

The first option excludes text nodes whose name starts with `#` which is a naming convention for text
nodes designed for app customization.

To use the second option, you will need to run the **Check/update keywords** plugin first to upload
the generated json file under your android app's build folder. The generated json file contains app
customizations for nodes so the localization plugin can exclude those text nodes from translation.

<img src="CheckOrUpdateKeywords.png">

> ### 2.2 Upload an strings xml

Reviewer 2: ```suggestion
### 2.2 Upload a strings xml file
```

Author: Done.

### 3.1 Review generated strings xml file

Now open Android Studio and let's start integrating the generated string xml file.

<img src="GenStringsExample.png">

The gen_strings.xml contains a string resource entry `label_hello`. If there are any conflicts with
string resources from other strings xml files, we recommend two options to resolve the conflicts:

1. Manually resolve the conflicts by renaming or removing the conflicted string resources in other
strings xml files.

2. Run the plugin and upload other string xml file containing conflicts to merge string resources
with the figma design file to avoid conflicts.

Next the strings can be sent for translations.

### 3.2 Update dcf file

Update the dcf file used by your app so it can look up the string resource when rendering.

### 3.3 Live update option

If you use live update, because DesignCompose uses the local string from the resource file, you will
not be able to see the changes you are making to the text nodes in live update. We have provided an option
to re-enable the live update for text nodes by unchecking the `Use local string resource if exists` option.

<img src="LiveUpdateLocalizationOption.png">

> ### 3.5 Localize app side customizations

Reviewer 2: You have 3.5 and 3.4 reversed

Author: Done.

<img src="LocalizationResults.png">

## 4. How the localization works behind the plugin

### 4.1 It uses plugin data

The localization plugin goes through all the text nodes of the figma design file and auto generates
a string resource name for each text node. The string resource name is saved to the shared plugin data.
When DesignCompose fetches the figma design file, it fetches the shared plugin data and later on the
string resource name will be used to look up the local string resource to render the text node.

Char limit and description are saved to private plugin data which will only be used to regenerate the
strings xml file and will not be fetched by DesignCompose.

The plugin also provides a **Clear String Resource** tool to clear the string resource names, char
limits and descriptions from the plugin data.

### 4.2 It supports mixed styles

For a text node with a single style, the plugin generates a string resource. And for a text node with
mixed styles, the plugin generates a string array resource with each item mapping to the text segments
with different styles.

## 5. What is next

We are working on a few improvements for the plugin:

> 1. the ability to exclude a text node manually before hitting the **Generate** button. Designers might

Reviewer 2: ```suggestion
1. The ability to exclude a text node manually before hitting the **Generate** button. Designers might
```

Author: Done.

add some text nodes as comments or explanations to design components which are not actually part of
the product design.

> 2. the ability to do different translations for the same text. Right now the plugin groups the text nodes

Reviewer 2: ```suggestion
2. The ability to do different translations for the same text. Right now the plugin groups the text nodes
```

Author: Done.

> with the same text to use the same string resource name. However, same text can mean differently under

Reviewer 2: ```suggestion
with the same text to use the same string resource name. However, the same text can have different translations under
```

Author: Done.

> different context.

Reviewer 2: ```suggestion
different contexts.
```

Author: Done.

## 6. Summary

> This plugin can not be 100% perfect to solve the localization difficulties, but we will

Reviewer 2: ```suggestion
This plugin does not perfectly solve all localization issues, but we will
```

Author: Done.

> continue to improve it to make localization easier for DesignCompose developers. Any suggestion

Reviewer 2: ```suggestion
continue to improve it to make localization easier for DesignCompose developers. Any suggestions
```

Author: Done.
